src/document/doc-utils.js

```diff
@@ -71,26 +71,51 @@ function traverseDoc(doc, onEnter, onExit, shouldTraverseConditionalGroups) {
 }
 
 function mapDoc(doc, cb) {
-  if (Array.isArray(doc)) {
-    return cb(doc.map((part) => mapDoc(part, cb)));
-  }
+  // Cache already processed docs in a map to make sure each doc is not
+  // processed more than once. It's important to do this because docs are
+  // often reused, especially in conditional groups.
```

Comment:
I don't understand why, isn't this function always return a new `Doc`? Unless the callback modify the original doc.

Reply:
You're right. It's just an optimization. I'll change the comment.

Reply:
But it's still important. Reused subtrees can be huge.